### PR TITLE
bpf: egressgw: sync logic to determine if destination is outside cluster 

### PR DESCRIPTION
**Pull request by jibi**
See individual commits

### CODE REVIEW

File: bpf/lib/nodeport.h [hunk 1]
```diff
@@ -1101,6 +1123,35 @@ static __always_inline bool snat_v4_needed(struct __ctx_buff *ctx, __be32 *addr,
 	 */
 	return false;
 # endif
+
+#if defined(ENABLE_EGRESS_GATEWAY)
+	/* If the packet is destined to an entity inside the cluster, either EP
+	 * or node, skip SNAT since only traffic leaving the cluster is supposed
+	 * to be masqueraded with an egress IP.
+	 */
```
Comment (kkourt) [inline bpf/lib/nodeport.h:1137]: It took me a while to realize (and it's obvious in retrospect) but what this patch does is to move the egress gw check _before_ the `IPV4_SNAT_EXCLUSION_DST_CIDR` check.

How about we add a comment here explaining why the egress gw check needs to happen before the `IPV4_SNAT_EXCLUSION_DST_CIDR`?

If we do so, maybe also change the commit message from:
> To fix this, this commit changes the snat_v4_needed() logic to match the
one we use in handle_ipv4_from_lxc().

To something like:
> To fix this, this commit changes the snat_v4_needed() logic to match the
one we use in handle_ipv4_from_lxc() and executes it before the  `IPV4_SNAT_EXCLUSION_DST_CIDR` check.

**Review by kkourt — APPROVED** [review]
Thanks Gilberto, LGTM.

As we discussed offline, I think we should describe what `IPV4_SNAT_EXCLUSION_DST_CIDR` and, subsequently, `--ipv4-native-routing-cidr` mean. My understanding was that the above CIDR included node IPs and only those, which at least in the EKS environment you tried out was not the case. IMO, this can be done in a followup PR but we should do it soon before it gets forgotten since there was a lot of time wasted on this based on different understandings of what above macro meant.

Based on our offline discussion `IPV4_SNAT_EXCLUSION_DST_CIDR` seems to be: "a CIDR that does not require SNAT (unless traffic is going through egress gw)". We should add an appropriate line in @pchaigno's list here: https://github.com/cilium/cilium/issues/15379#issuecomment-897441189.

File: bpf/lib/nodeport.h (hunk 1 above)
Comment (jibi) [reply bpf/lib/nodeport.h:1137]: Done :heavy_check_mark:

**Comment by jibi** [comment]
> Thanks Gilberto, LGTM.
> 
> As we discussed offline, I think we should describe what `IPV4_SNAT_EXCLUSION_DST_CIDR` and, subsequently, `--ipv4-native-routing-cidr` mean. My understanding was that the above CIDR included node IPs and only those, which at least in the EKS environment you tried out was not the case. IMO, this can be done in a followup PR but we should do it soon before it gets forgotten since there was a lot of time wasted on this based on different understandings of what above macro meant.
> 
> Based on our offline discussion `IPV4_SNAT_EXCLUSION_DST_CIDR` seems to be: "a CIDR that does not require SNAT (unless traffic is going through egress gw)". We should add an appropriate line in @pchaigno's list here: [#15379 (comment)](https://github.com/cilium/cilium/issues/15379#issuecomment-897441189).

Yep, I'll fill a PR with this changes now :+1:

**Comment by jibi** [comment]
Marking as ready to merge. No need to rerun full CI as I just added an additional comment and amended a commit message (and before CI was green)

**Comment by jibi** [comment]
(taking care of v1.10 backport as there may be a few conflicts here)

**Pull request merged**